File: Problem162.cpp
/*162. Find Peak Element
A peak element is an element that is greater than its neighbors.*/

#include <iostream>
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

int findPeakElement(vector<int>& nums) {
    int ind = -1;
    if (nums.size()>1){
        if (nums[0] > nums[1]) ind = 0;
        for (int i=1; i < nums.size()-1; i++){
            if ((nums[i] > nums[i-1])&&(nums[i] > nums[i+1])) {
                ind = i;
                break;
            }
        }
        if (nums[nums.size()-1] > nums[nums.size()-2]) ind = nums.size()-1;
    } else {
        ind = 0;
    }
    return ind;
}

//Binary Search
//int findPeakElement(const vector<int> &num) {
//    return Helper(num, 0, num.size()-1);
//}
//int Helper(const vector<int> &num, int low, int high)
//{
//    if(low == high)
//        return low;
//    else
//    {
//        int mid1 = (low+high)/2;
//        int mid2 = mid1+1;
//        if(num[mid1] > num[mid2])
//            return Helper(num, low, mid1);
//        else
//            return Helper(num, mid2, high);
//    }
//  }
//};

int main() {
    vector<int> v = {1,2,3,1};
    int i = findPeakElement(v);
    cout << i << endl;
//    for (int i = 0; i < v.size(); i++) {
//        cout<<v[i]<< " ";
//    }
    return 0;
}

File: NewJob2025/Problem162.java
// 162. Find Peak Element - M
    public int findPeakElement(int[] nums) {
        if(nums.length == 1) {return 0;}

        int low = 0; int high = nums.length-1;
        int mid;
        while (low <= high) {
            mid = low + (high-low) / 2;
            // edge
            if (mid == nums.length-1) {
                if (nums[mid] > nums[mid-1]) return mid;
                high = mid - 1;
                continue;
            }

            if (mid == 0) {
                if (nums[mid] > nums[mid+1]) return mid;
                low = mid + 1;
                continue;
            }

            if ((nums[mid] > nums[mid+1]) && (nums[mid] > nums[mid-1])) {
                return mid;
            } 

            if ((nums[mid] > nums[mid+1]) && (nums[mid] < nums[mid-1])) {
                high = mid - 1;
                continue;
            } else {
                low = mid+1;
                continue;
            }

        }

        return -1;
        
    }

// could be simplified
            if(nums[mid] > nums[mid-1] && nums[mid] > nums[mid+1]) return mid;
            else if(nums[mid] < nums[mid-1]) end = mid - 1;
            else if(nums[mid] < nums[mid+1]) start = mid + 1;
